### PR TITLE
Feature/#16 map missing information

### PR DESCRIPTION
**Pull request by LoyVanBeek**
Solves #16

### CODE REVIEW

**Comment by MatthijsBurgh** [comment]
Who should review this?

**Comment by alberth** [comment]
Defining the set of entry-points of the grammar outside the grammar file looks quite fragile to say the least. All you need really is a known set of targets that the engine can just use.

To nail stuff down, that set should be defined in the interface to whatever thing converts the request to an answer. More practically, a set of constants in a message definition.

The 'thing' then needs to be extended to deal with this. The simplest approach is to just forward the target of the message  to the grammar parser, which then has to do connect it to a grammar rule. The simplest solution there is to have a hard-coded mapping of requested value to rule-name, which can be checked eg for existence in the grammar. More fancy grammar parsers can give the user a way to specify the mapping in the file, although it's likely not needed.

**Comment by LoyVanBeek** [comment]
I admit it does feel kinda weird defining this mapping here. 

But the grammar_pars doesn't/should not know anything about the action_server, nor does & should the action_server know anything about the grammar_parser.

Th conversation_engine is the link between these decoupled systems. That means setting up the link between the actions and the grammar here.

**Comment by alberth** [comment]
> But the grammar_pars doesn't/should not know anything about the action_server, nor does & should the action_server know anything about the grammar_parser.

Grammar not knowing about the application is a myth. It has to provide well-known targets or the conversation engine fails. See https://github.com/tue-robotics/conversation_engine/pull/17/files#diff-087bb20ab859fc2e9166f0b131bd1a41R510 which fails hard if a needed target is not found.

Grammar not knowing about the action server is also myth. It produces a json data structure that has to magically match expectations of the action server, or the action server fails to function.

Right now, all this knowledge is scattered at many points in the code, and you maintain a whole additional layer to connect all the various parts and make it work. By explicitly defining names and expectations etc you can refer to such names and concepts in the code without that layer.

Theoretically, this is less flexible, but since new functionality involves a lot of changes in code etc, defining an additional entry and name doesn't seem to be such a major problem to me.

**Comment by LoyVanBeek** [comment]
> Grammar not knowing about the application is a myth. It has to provide well-known targets or the conversation engine fails.

True, the grammar is specifically tailored for the action_server, to produce dictionary the action_server understands. The grammar **_parser** is indifferent to the action_server but because of the grammar, it produces something useful. 

I'd like to keep the conversation_engine itself free of knowledge of a specif grammar (although we probably never use it with a very different grammar).

File: scripts/engine [hunk 1]
```diff
@@ -12,6 +12,17 @@ def main():
     challenge = rospy.get_param("~challenge",
                                 None)  # Challenge to load grammar and command_target from if `grammar` and `command_target` are empty
 
+    # Which grammar target to use for what field of missing information?
+    grammar_target_for_missing_info = rospy.get_param("~grammar_target_for_missing_info", {})
+
+    grammar_target_for_missing_info = {
```
Comment (alberth) [inline scripts/engine]: Getting a value from the parameter server and then immediately overwriting it with a default(?) looks like the wrong way around. Perhaps swap both assignments, and use the variable as default in the parameter server retrieval call as default value?

Comment (alberth) [reply scripts/engine]: was already fixed I see. sorry

**Comment by MatthijsBurgh** [comment]
If this should be merged, it should be merged into master not into robocup. Therefore closing this PR.

**Pull request closed without merging**